alacritty/src/event.rs

```diff
@@ -64,6 +65,386 @@ const MAX_SEARCH_HISTORY_SIZE: usize = 255;
+                    window_context.dirty = true;
+                    if window_context.display.window.has_frame {
+                        window_context.display.window.request_redraw();
+                    }
+                }
+            },
+            EventType::Terminal(TerminalEvent::Exit) => {
+                // Remove the closed terminal.
+                let window_context = match self.windows.remove(&window_id) {
+                    Some(window_context) => window_context,
+                    None => return,
+                };
+
+                // Unschedule pending events.
+                self.scheduler.unschedule_window(window_context.id());
+
+                // Shutdown if no more terminals are open.
+                if self.windows.is_empty() {
+                    // Write ref tests of last window to disk.
+                    if self.config.debug.ref_test {
+                        window_context.write_ref_test_results();
+                    }
+
+                    event_loop.exit();
+                }
+            },
+            // NOTE: This event bypasses batching to minimize input latency.
+            EventType::Frame => {
+                if let Some(window_context) = self.windows.get_mut(&window_id) {
+                    window_context.display.window.has_frame = true;
+                    if window_context.dirty {
+                        window_context.display.window.request_redraw();
+                    }
+                }
+            },
+            _ => {
+                if let Some(window_context) = self.windows.get_mut(&window_id) {
+                    window_context.handle_event(
+                        #[cfg(target_os = "macos")]
+                        event_loop,
+                        &self.proxy,
+                        &mut self.clipboard,
+                        &mut self.scheduler,
+                        WinitEvent::UserEvent(event),
+                    );
+                }
+            },
+        }
+    }
+
+    fn about_to_wait(&mut self, event_loop: &ActiveEventLoop) {
+        if self.config.debug.print_events {
+            info!("About to wait");
+        }
+
+        // Dispatch event to all windows.
+        for window_context in self.windows.values_mut() {
+            window_context.handle_event(
+                #[cfg(target_os = "macos")]
+                event_loop,
+                &self.proxy,
+                &mut self.clipboard,
+                &mut self.scheduler,
+                WinitEvent::AboutToWait,
+            );
+        }
+
+        // Update the scheduler after event processing to ensure
+        // the event loop deadline is as accurate as possible.
+        let control_flow = match self.scheduler.update() {
+            Some(instant) => ControlFlow::WaitUntil(instant),
+            None => ControlFlow::Wait,
+        };
+        event_loop.set_control_flow(control_flow);
+    }
+
+    fn exiting(&mut self, _event_loop: &ActiveEventLoop) {
```

Comment:
Couldn't the body of this function just be a drop impl on `self`? I don't see which parts of this require being called manually before exiting the event loop.

Shouldn't reinvent destructors.

Reply:
The `self` is `&mut` borrowed and not moved into the loop, so you can not `move` self, you can probably move `event_loop` here in though, but that's about it?

Reply:
Things will get dropped after this anyway though I'd assume? What's the necessity of doing it in `exiting` rather than later down the line?

Reply:
I mean, maybe you have something around in your state you don't want to drop? Passing the loop in also doesn't work with APIs which don't destroy the loop after exit.

Reply:
Just to make sure I understand this correctly: Both the EGL display AND the clipboard **must** be dropped before the event loop, because otherwise things explode?

And event loop shutdown automatically drops the event loop, so you can't take a return value from `event_loop.run` or whatever and manually drop that later?

Reply:
The must be dropped because they _borrow_ the resource, the display pointer, which is getting destroyed with the loop. There's no way to prevent this sort of things in general, because no way to indicate that to event loop.

Returning the event loop back out of run and making it invalid for further reruns is probably good, but it'll still require the client to keep things _sound_.

Reply:
Couldn't the event loop just not destroy the display pointer?

Reply:
And leak the resource? It could, I guess, but there're plans to just ref-count it.